src/rgw/driver/rados/rgw_notify.cc

```diff
@@ -779,24 +814,31 @@ constexpr uint32_t WORKER_COUNT = 1;                 // 1 worker thread
 constexpr uint32_t STALE_RESERVATIONS_PERIOD_S = 120;   // cleanup reservations that are more than 2 minutes old
 constexpr uint32_t RESERVATIONS_CLEANUP_PERIOD_S = 30; // reservation cleanup every 30 seconds
 
-bool init(CephContext* cct, rgw::sal::RadosStore* store,
-          const SiteConfig& site, const DoutPrefixProvider *dpp) {
+bool init(const DoutPrefixProvider* dpp, rgw::sal::RadosStore* store,
+          const SiteConfig& site) {
   if (s_manager) {
+    ldpp_dout(dpp, 1) << "ERROR: failed to init notification manager: already exists" << dendl;
+    return false;
+  }
+  if (!RGWPubSubEndpoint::init_all(dpp->get_cct())) {
     return false;
   }
   // TODO: take conf from CephContext
-  s_manager = new Manager(cct, MAX_QUEUE_SIZE, 
+  s_manager = std::make_unique<Manager>(dpp->get_cct(), MAX_QUEUE_SIZE, 
       Q_LIST_UPDATE_MSEC, Q_LIST_RETRY_MSEC, 
       IDLE_TIMEOUT_USEC, FAILOVER_TIME_MSEC, 
       STALE_RESERVATIONS_PERIOD_S, RESERVATIONS_CLEANUP_PERIOD_S,
       WORKER_COUNT,
       store, site);
+  s_manager->init();
   return true;
 }
 
 void shutdown() {
-  delete s_manager;
-  s_manager = nullptr;
+  if (!s_manager) return;
+  RGWPubSubEndpoint::shutdown_all();
```

Comment:
if the endpoints are all stopped, there is nothing to invoke the callbacks for the completions.

Reply:
if outstanding completions aren't canceled, i would expect `process_entry()`'s calls to `push_endpoint->send()` to hang and prevent the io_context threads from joining

separately, what happens if we call `push_endpoint->send()` after `RGWPubSubEndpoint::shutdown_all()`?

Reply:
> if outstanding completions aren't canceled, i would expect `process_entry()`'s calls to `push_endpoint->send()` to hang and prevent the io_context threads from joining
> 

* in case of kafka, we call all completions with an erro code:
https://github.com/ceph/ceph/blob/main/src/rgw/rgw_kafka.cc#L101
* similar handling in case of amqp:
https://github.com/ceph/ceph/blob/main/src/rgw/rgw_amqp.cc#L146
* for http, i call "stop()" and then dtor:
https://github.com/ceph/ceph/pull/56979/files#diff-a7d6265d00f7a11ffed731f280898bebf406a80b3ccdcecc50c13c48903d669eR416
this is similar to the global "rgw_http_client_cleanup()".
however, i don't see any code there that calcel any pending work.
@cbodley do you think that calling "RGWHTTPManager::manage_pending_requests()" would fix that?

> separately, what happens if we call `push_endpoint->send()` after `RGWPubSubEndpoint::shutdown_all()`?

the "manager" pointers should be deleted and set to null. this should return an "-ESRCH" error.  
however, i do see this crash:
```
     "/lib64/libc.so.6(+0x3e6f0) [0x7fb79c23e6f0]",
        "__pthread_rwlock_wrlock()",
        "(RGWHTTPManager::register_request(rgw_http_req_data*)+0x37) [0x55a1f56c76e7]",
        "(RGWHTTPManager::add_request(RGWHTTPClient*)+0x168) [0x55a1f56c8c48]",
        "radosgw(+0x80ab13) [0x55a1f58b3b13]",
        "radosgw(+0x80f52b) [0x55a1f58b852b]",
        "(void boost::context::detail::context_entry<boost::context::detail::record<boost::context::continuation, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits>, spawn::detail::spawn_helper<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor>, rgw::notify::Manager::process_queue(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >)::{lambda(spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >)#7}, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits> >::operator()()::{lambda(boost::context::continuation&&)#1}> >(boost::context::detail::transfer_t)+0x55) [0x55a1f58b92f5]",
        "make_fcontext()"
```
probably a race condition with the deletion of the pointer.
will protect it with a shared mutex, so that the different sender don't block one another, and only the shutdown

Reply:
> however, i don't see any code there that calcel any pending work.
> @cbodley do you think that calling "RGWHTTPManager::manage_pending_requests()" would fix that?

it looks like `RGWHTTPManager::reqs_thread_entry()`, below the `while (!going_down)` loop, does loop over outstanding requests to call `_unlink_request()` -> `_finish_request(req_data, -ECANCELED);`